### PR TITLE
C++/Python getTraitProperty consistency.

### LINKED ISSUE #57: Consistency between `C++` and `Python`
## What

Ensure C++ and Python behave consistently for missing trait/properties.

## ACs

- Determine inconsistencies and define resulting behaviour
- Trait-gen generated code should have the same failure modes in C++ and Python
- #45 done to ensure coverage of defined behaviours in both languages.

### PR DESCRIPTION
**Pull request by elliotcmorris**
Closes #57 

C++ `getTraitProperty` method now returns `std::optional<T>` rather than throwing when invoked without a default parameter. This mirrors the python behaviour of returning `None` in this case.

Also adds pair of tests to python test suite to add missing checks about traitViews constructed around empty traitsdatas.

### CODE REVIEW

**Comment by elliotcmorris** [comment]
Following on from our discussion, didn't spot anything on a squiz through of `C_API`, `CODING_STYLE` and `CODING_STANDARDS` about not using standard library types in an interface because of the c-api.

File: python/openassetio_traitgen/templates/cpp/trait.hpp.in [hunk 1]
```diff
@@ -109,19 +110,19 @@ public:
...
-   * not found or is of an unexpected type.
+   * Gets the value of the {{ property.id }} property. Returns an empty
+   * optional if not found or is of an unexpected type.
    *
    * {{ property.description |  wordwrap(69, wrapstring="\n* ") | indent(4) }}
    */
-  [[nodiscard]] {{ VarType }} get{{ VarMethodName }}() const {
-    if (property::Value value; traitsData_->getTraitProperty(&value, kId, "{{ property.id }}")) {
+  [[nodiscard]] std::optional<{{ VarType }}> get{{ VarMethodName }}() const {
+  if (property::Value value; traitsData_->getTraitProperty(&value, kId, "{{ property.id }}")) {
```
Comment (feltech) [inline python/openassetio_traitgen/templates/cpp/trait.hpp.in:119]: [fix] Lost some indentation here that makes the generated output a bit weird.

**Review by feltech — COMMENTED** [review]
Some minor comments, will finish review asap.

File: tests/generators/cpp/src/test.cpp [hunk 2]
```diff
@@ -347,10 +354,7 @@ TEMPLATE_TEST_CASE("Property getters", "", openassetio_abi::Bool, openassetio_ab
     const Fixture fixture{AllPropertiesTrait{traitsData}};
 
     WHEN("property is queried without a default") {
-      THEN("exception is thrown") {
-        CHECK_THROWS_MATCHES(fixture.getProperty(), std::runtime_error,
-                             Catch::Matchers::Message("Property is not set."));
-      }
+      THEN("optional return is empty") { CHECK(!fixture.getProperty().has_value()); }
```
[the remark this answers is not in the document]
Comment (elliotcmorris) [reply tests/generators/cpp/src/test.cpp:357]: You mean get the value and store it in the WHEN, then check it in the THEN?

File: tests/generators/test_cpp.py [hunk 3]
```diff
@@ -181,8 +181,8 @@ def test_has_prefixed_property_getters_with_default_with_expected_docstring(
             )
             == f"""
   /**
-   * Gets the value of the {property_name} property or throws if
-   * not found or is of an unexpected type.
+   * Gets the value of the {property_name} property. Returns an empty
```
[the remark this answers is not in the document]
Comment (elliotcmorris) [reply tests/generators/test_cpp.py:184]: "empty optional" is used in method documentation on that very same page, so I don't think there's an argument for preferred terminology.

File: python/openassetio_traitgen/templates/cpp/trait.hpp.in [hunk 4]
```diff
@@ -109,19 +110,19 @@ public:
...
       if ({{ VarType }}* maybeOut = std::get_if<{{ VarType }}>(&value)) {
         return *maybeOut;
       }
       throw std::runtime_error{"Invalid stored value type: should be '{{ VarType }}'."};
     }
-    throw std::runtime_error{"Property is not set."};
+    return std::optional<{{VarType}}>{};
```
[the remark this answers is not in the document]
Comment (elliotcmorris) [reply python/openassetio_traitgen/templates/cpp/trait.hpp.in:125]: There's no real difference other than expression right? I'm happy as it stands.

File: RELEASE_NOTES.md [hunk 5]
```diff
@@ -1,6 +1,16 @@
 Release Notes
 =============
 
+v1.0.0-alpha.x
+--------------
+
+### Breaking Changes
+
+- C++ `getTraitProperty` method now returns `std::optional<T>` rather
```
[the remark this answers is not in the document]
Comment (elliotcmorris) [reply RELEASE_NOTES.md]: Aye good shout. Wasn't sure about the angle brackets to convey the different type methods in that way though, just left them out. https://github.com/OpenAssetIO/OpenAssetIO-TraitGen/pull/63/commits/eca2963e0ea3ff02ff6eae02f79da9bfa5eca5ea

File: python/openassetio_traitgen/templates/cpp/trait.hpp.in (hunk 1 above)
Comment (elliotcmorris) [reply python/openassetio_traitgen/templates/cpp/trait.hpp.in]: https://github.com/OpenAssetIO/OpenAssetIO-TraitGen/pull/63/commits/03b3c1df60455e678dbe8a454e0b2c467abe10f4

File: tests/generators/test_cpp.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (feltech) [reply tests/generators/test_cpp.py:184]: :open_mouth: missed that, fair enough

File: tests/generators/cpp/src/test.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (feltech) [reply tests/generators/cpp/src/test.cpp:357]: Yeah, like I say super minor, mostly for neatness, but does have _some_ utility, e.g. if a failure is in the WHEN it'll be reported independently.

File: RELEASE_NOTES.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (feltech) [reply RELEASE_NOTES.md]: Also similar comment for PR description

**Comment by feltech** [comment]
The first commit `[Core] Copy python error mode in getTraitProperty` could do with updating with the issue number and `getTraitProperty` -> `getSomeProperty` or similar.

File: tests/generators/test_python.py [hunk 6]
```diff
@@ -448,6 +448,22 @@ def test_when_property_not_set_then_returns_None(
 
         assert getter() is None
 
+    def test_when_get_property_on_blank_TraitsData_view_then_returns_None(
```
Comment (foundrytom) [inline tests/generators/test_python.py:451]: [consider][strong] Be a little clearer about what exactly the when is, and match the convention for the other cases (same below):
```suggestion
    def test_when_trait_not_set_then_returns_None(
```

**Review by foundrytom — APPROVED** [review]

File: tests/generators/cpp/src/test.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (elliotcmorris) [reply tests/generators/cpp/src/test.cpp:357]: I see. Guess i'm gonna leave this as it is becuase the scope of this change extends beyond the altered methods, as this sort of pattern is done elsewhere as well, and as you say, super minor.

File: tests/generators/test_python.py (hunk 6 above)
Comment (elliotcmorris) [reply tests/generators/test_python.py]: Sure, was trying to mirror the given/when wording from the cpp test, but had to munge it a fair bit to get the longer default one to fit inside the line length. Have taken your suggestion instead. 
https://github.com/OpenAssetIO/OpenAssetIO-TraitGen/pull/63/commits/ccdeea88fde9b4ba90c1105bc93b40aa5dbfdc32

**Comment by elliotcmorris** [comment]
@feltech Changed the first commit in a rebase as suggested.

File: tests/generators/test_python.py (hunk 6 above)
Comment (foundrytom) [reply tests/generators/test_python.py]: Thanks - for some reason that diff is mind boggling, but the end result looks 👌

**Review by feltech — APPROVED** [review]
